In [1]:
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np # linear algebra

- Fontes:

https://www.kaggle.com/mylesoneill/magic-the-gathering-cards?select=PioneerCards.json

https://mtgjson.com/downloads/all-files/#pioneer

In [2]:
# Importando o arquivo
cards = pd.read_json('Arquivos/PioneerCards.json')

In [3]:
# Listando as 5 primeiras linhas
cards.head()

Abandon Reason  \
colorIdentity                                                    [R]   
colors                                                           [R]   
convertedManaCost                                                  3   
edhrecRank                                                      7558   
foreignData        [{'language': 'German', 'name': 'Abkehr von de...   

                                               Abandoned Sarcophagus  \
colorIdentity                                                     []   
colors                                                            []   
convertedManaCost                                                  3   
edhrecRank                                                      6972   
foreignData        [{'flavorText': 'Was einst ruhte, ruht nicht m...   

                                                 Abbot of Keral Keep  \
colorIdentity                                                    [R]   
colors                                                           [R]   
convertedManaCost                                                  2   
edhrecRank                                                      6887   
foreignData        [{'language': 'German', 'name': 'Abt des Keral...   

                                                 Aberrant Researcher  \
colorIdentity                                                    [U]   
colors                                                           [U]   
convertedManaCost                                                  4   
edhrecRank                                                     14063   
foreignData        [{'flavorText': '„Die Metamorphose ist ein and...   

                                                  Abhorrent Overlord  \
colorIdentity                                                    [B]   
colors                                                           [B]   
convertedManaCost                                                  7   
edhrecRank                                                      1511   
foreignData        [{'language': 'German', 'name': 'Verhasster Ob...   

                                                  Abnormal Endurance  \
colorIdentity                                                    [B]   
colors                                                           [B]   
convertedManaCost                                                  2   
edhrecRank                                                      5999   
foreignData        [{'flavorText': 'Die Hoffnung stirbt zuletzt, ...   

                                             Abolisher of Bloodlines  \
colorIdentity                                                    [B]   
colors                                                            []   
convertedManaCost                                                  5   
edhrecRank                                                      2481   
foreignData        [{'flavorText': 'Selbst wenn es Blut regnen wü...   

                                                Abomination of Gudul  \
colorIdentity                                              [B, G, U]   
colors                                                     [B, G, U]   
convertedManaCost                                                  6   
edhrecRank                                                     13679   
foreignData        [{'language': 'German', 'name': 'Gudul-Scheußl...   

                                                              Abrade  \
colorIdentity                                                    [R]   
colors                                                           [R]   
convertedManaCost                                                  2   
edhrecRank                                                      1219   
foreignData        [{'flavorText': 'Die Wüste ist eine hungrige B...   

                                                        Abrupt Decay  ...  \
colorIdentity                                                 [B, G]  ...   
col

In [4]:
# Transpondo o DataFrame: Escrevendo as colunas como linhas e vice-versa.
cards = cards.T

In [5]:
# Listando as 5 primeiras linhas com o novo formato.
cards.head()

colorIdentity colors convertedManaCost edhrecRank  \
Abandon Reason                  [R]    [R]                 3       7558   
Abandoned Sarcophagus            []     []                 3       6972   
Abbot of Keral Keep             [R]    [R]                 2       6887   
Aberrant Researcher             [U]    [U]                 4      14063   
Abhorrent Overlord              [B]    [B]                 7       1511   

                                                             foreignData  \
Abandon Reason         [{'language': 'German', 'name': 'Abkehr von de...   
Abandoned Sarcophagus  [{'flavorText': 'Was einst ruhte, ruht nicht m...   
Abbot of Keral Keep    [{'language': 'German', 'name': 'Abt des Keral...   
Aberrant Researcher    [{'flavorText': '„Die Metamorphose ist ein and...   
Abhorrent Overlord     [{'language': 'German', 'name': 'Verhasster Ob...   

                          layout  \
Abandon Reason            normal   
Abandoned Sarcophagus     normal   
Abbot of Keral Keep       normal   
Aberrant Researcher    transform   
Abhorrent Overlord        normal   

                                                              legalities  \
Abandon Reason         {'commander': 'Legal', 'duel': 'Legal', 'legac...   
Abandoned Sarcophagus  {'commander': 'Legal', 'duel': 'Legal', 'legac...   
Abbot of Keral Keep    {'commander': 'Legal', 'duel': 'Legal', 'legac...   
Aberrant Researcher    {'commander': 'Legal', 'duel': 'Legal', 'legac...   
Abhorrent Overlord     {'commander': 'Legal', 'duel': 'Legal', 'legac...   

                        manaCost mtgoFoilId mtgoId  ... mtgArenaId power  \
Abandon Reason            {2}{R}      61409  61408  ...        NaN   NaN   
Abandoned Sarcophagus        {3}      64841  64840  ...      65833   NaN   
Abbot of Keral Keep       {1}{R}      58043  58042  ...        NaN     2   
Aberrant Researcher       {3}{U}      59887  59886  ...        NaN     3   
Abhorrent Overlord     {5}{B}{B}      50401  50400  ...        NaN     6   

                      toughness faceConvertedManaCost  \
Abandon Reason              NaN                   NaN   
Abandoned Sarcophagus       NaN                   NaN   
Abbot of Keral Keep           1                   NaN   
Aberrant Researcher           2                     4   
Abhorrent Overlord            6                   NaN   

                                                       names side  \
Abandon Reason                                           NaN  NaN   
Abandoned Sarcophagus                                    NaN  NaN   
Abbot of Keral Keep                                      NaN  NaN   
Aberrant Researcher    [Aberrant Researcher, Perfected Form]    a   
Abhorrent Overlord                                        []  NaN   

                      leadershipSkills loyalty colorIndicator hasNoDeckLimit  
Abandon Reason                     NaN     NaN            NaN            NaN  
Abandoned Sarcophagus              NaN     NaN            NaN            NaN  
Abbot of Keral Keep                NaN     NaN            NaN            NaN  
Aberrant Researcher                NaN     NaN            NaN            NaN  
Abhorrent Overlord                 NaN     NaN            NaN            NaN  

[5 rows x 31 columns]

In [6]:
# Criando uma nova coluna para ser o meu index
cards['index'] = range(0,cards.shape[0])

In [7]:
# Setando a nova coluna de indice
cards = cards.set_index('index')

In [8]:
# Retornando 5 valores aleatórios do dataframe
cards.sample(5)

colorIdentity colors convertedManaCost edhrecRank  \
index                                                     
666              []     []                 1      13688   
5197             []     []                 3      15745   
5735            [U]     []                 4      11805   
1160            [B]    [B]                 7      14268   
5647            [W]    [W]                 2       9313   

                                             foreignData  layout  \
index                                                              
666    [{'flavorText': 'Treu ergeben befolgt es jeden...  normal   
5197   [{'flavorText': '„Hör erst auf zu hacken, wenn...  normal   
5735   [{'language': 'German', 'name': 'Tezzerets Tor...  normal   
1160   [{'language': 'German', 'name': 'Blutpakt', 't...  normal   
5647   [{'flavorText': '„Dein Tod ist Balsam für die ...  normal   

                                              legalities manaCost mtgoFoilId  \
index                                                                          
666    {'commander': 'Legal', 'duel': 'Legal', 'legac...      {1}      58113   
5197   {'commander': 'Legal', 'duel': 'Legal', 'legac...      {3}      61489   
5735   {'commander': 'Legal', 'duel': 'Legal', 'histo...      {4}        NaN   
1160   {'commander': 'Legal', 'duel': 'Legal', 'legac...   {6}{B}      53687   
5647   {'brawl': 'Legal', 'commander': 'Legal', 'duel...   {1}{W}        NaN   

      mtgoId  ... mtgArenaId power toughness faceConvertedManaCost names side  \
index         ...                                                               
666    58112  ...        NaN     2         1                   NaN   NaN  NaN   
5197   61488  ...        NaN   NaN       NaN                   NaN   NaN  NaN   
5735   68111  ...      68260   NaN       NaN                   NaN   NaN  NaN   
1160   53686  ...        NaN   NaN       NaN                   NaN   NaN  NaN   
5647   68241  ...      67760   NaN       NaN                   NaN   NaN  NaN   

      leadershipSkills loyalty colorIndicator hasNoDeckLimit  
index                                                         
666                NaN     NaN            NaN            NaN  
5197               NaN     NaN            NaN            NaN  
5735               NaN     NaN            NaN            NaN  
1160               NaN     NaN            NaN            NaN  
5647               NaN     NaN            NaN            NaN  

[5 rows x 31 columns]

In [9]:
# Retornando o nome das colunas
cards.columns

Index(['colorIdentity', 'colors', 'convertedManaCost', 'edhrecRank',
       'foreignData', 'layout', 'legalities', 'manaCost', 'mtgoFoilId',
       'mtgoId', 'name', 'printings', 'purchaseUrls', 'rulings',
       'scryfallOracleId', 'subtypes', 'supertypes', 'text', 'type', 'types',
       'uuid', 'mtgArenaId', 'power', 'toughness', 'faceConvertedManaCost',
       'names', 'side', 'leadershipSkills', 'loyalty', 'colorIndicator',
       'hasNoDeckLimit'],
      dtype='object')

In [10]:
# Retornando a quantidade de linhas nas colunas
cards.count()

colorIdentity            6567
colors                   6567
convertedManaCost        6567
edhrecRank               6553
foreignData              6567
layout                   6567
legalities               6567
manaCost                 6283
mtgoFoilId               3687
mtgoId                   5419
name                     6567
printings                6567
purchaseUrls             5961
rulings                  6567
scryfallOracleId         6567
subtypes                 6567
supertypes               6567
text                     6435
type                     6567
types                    6567
uuid                     6567
mtgArenaId               2661
power                    3579
toughness                3579
faceConvertedManaCost     321
names                     853
side                      330
leadershipSkills          429
loyalty                   150
colorIndicator             37
hasNoDeckLimit              3
dtype: int64

In [11]:
# Verificando a quantidade de linhas x colunas do Dataframe
cards.shape

(6567, 31)

In [12]:
# Retornando informações estatísticas do dataframe
cards.describe()

colorIdentity colors  convertedManaCost  edhrecRank foreignData  \
count           6567   6567             6567.0        6553        6567   
unique            27     35               14.0        6395        5794   
top              [B]    [W]                3.0       15329          []   
freq             980    989             1525.0           2         774   

        layout                                         legalities manaCost  \
count     6567                                               6567     6283   
unique       7                                                 28      381   
top     normal  {'commander': 'Legal', 'duel': 'Legal', 'legac...   {1}{W}   
freq      6223                                               1885      238   

        mtgoFoilId  mtgoId  ... mtgArenaId power toughness  \
count         3687    5419  ...       2661  3579      3579   
unique        3629    5301  ...       2605    17        17   
top          61133   61462  ...      70470     2         2   
freq             2       2  ...          2  1129       925   

       faceConvertedManaCost names side  \
count                  321.0   853  330   
unique                   8.0   165    3   
top                      2.0    []    b   
freq                    72.0   523  164   

                                         leadershipSkills loyalty  \
count                                                 429     150   
unique                                                  2       7   
top     {'brawl': False, 'commander': True, 'oathbreak...       5   
freq                                                  278      59   

       colorIndicator hasNoDeckLimit  
count              37              3  
unique              7              1  
top               [R]           True  
freq               10              3  

[4 rows x 31 columns]

In [13]:
# Retornando informações sobre as colunas, os dtypes, uso de memória
cards.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6567 entries, 0 to 6566
Data columns (total 31 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   colorIdentity          6567 non-null   object
 1   colors                 6567 non-null   object
 2   convertedManaCost      6567 non-null   object
 3   edhrecRank             6553 non-null   object
 4   foreignData            6567 non-null   object
 5   layout                 6567 non-null   object
 6   legalities             6567 non-null   object
 7   manaCost               6283 non-null   object
 8   mtgoFoilId             3687 non-null   object
 9   mtgoId                 5419 non-null   object
 10  name                   6567 non-null   object
 11  printings              6567 non-null   object
 12  purchaseUrls           5961 non-null   object
 13  rulings                6567 non-null   object
 14  scryfallOracleId       6567 non-null   object
 15  subtypes             

## Tratando o dataframe
- Remover as colunas que não serão necessárias para a nossa análise.
- Tratando os valores nulos
- Há valores do tipo Lista dentro do Dataframe, precisamos tratá-los também.
    - Colunas: colors, types, subtypes, printings, manaCost, supertypes

**Fonte Descrição dos Campos:**
- https://mtgjson.com/file-models/card-atomic/

In [14]:
# Removendo colunas que não serão utilizadas para a análise seguindo a documentação
colunasRemovidas = ['hasNoDeckLimit', 'colorIndicator', 'leadershipSkills', 'names', 'mtgArenaId', 'uuid', 'text',
                   'scryfallOracleId', 'rulings', 'purchaseUrls', 'mtgoId', 'mtgoFoilId', 'legalities', 'foreignData',
                   'edhrecRank', 'colorIdentity', 'faceConvertedManaCost','type', 'side', 'printings']

#Dropando as colunas e atualizando em memória
cards.drop(colunasRemovidas, axis=1, inplace=True)

In [15]:
# Listando o novo dataframe somente com as colunas que serão analisadas.
cards.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6567 entries, 0 to 6566
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   colors             6567 non-null   object
 1   convertedManaCost  6567 non-null   object
 2   layout             6567 non-null   object
 3   manaCost           6283 non-null   object
 4   name               6567 non-null   object
 5   subtypes           6567 non-null   object
 6   supertypes         6567 non-null   object
 7   types              6567 non-null   object
 8   power              3579 non-null   object
 9   toughness          3579 non-null   object
 10  loyalty            150 non-null    object
dtypes: object(11)
memory usage: 615.7+ KB


In [16]:
#Listando todas as colunas que possuem valores nulos.
cards.isnull().sum()

colors                  0
convertedManaCost       0
layout                  0
manaCost              284
name                    0
subtypes                0
supertypes              0
types                   0
power                2988
toughness            2988
loyalty              6417
dtype: int64

- A coluna manaCost possui **284** valores nulos
- A coluna power possui **2988** valores nulos
- A coluna toughness possui **2988** valores nulos
- A coluna loyalty possui **6417** valores nulos


In [17]:
# Tratando os valores nulos - ManaCost
# Quando não há custo de mana, considera-se que a carta não possui custo
cards['manaCost'].fillna(0, inplace=True)

In [18]:
# Tratando valores nulos - power
# Sugestão: Fazer uma análise mais detalhada para verificar se as cartas que não registram poder são Sorcery ou instants
cards['power'].fillna(0, inplace=True)

In [19]:
# Tratando valores nulos - toughness
# Sugestão: Fazer uma análise mais detalhada para verificar se as cartas que não registram toughness são Sorcery ou instants
cards['toughness'].fillna(0, inplace=True)

In [20]:
# Tratando valores nulos - loyalty
# Sugestão: Verificar se apenas cartas PW possuem loyalty
# PW não podem vim com loyalty 0
cards['loyalty'].fillna(0, inplace=True)

In [21]:
# Não há mais nenhum valor nulo
cards.isnull().sum()

colors               0
convertedManaCost    0
layout               0
manaCost             0
name                 0
subtypes             0
supertypes           0
types                0
power                0
toughness            0
loyalty              0
dtype: int64

In [22]:
# Convertendo a coluna types do tipo lista para Strings
# sugestão: Destrinchar isso aqui para melhor entendimento.
cardsTypes = cards['types'].apply(pd.Series)
cardsTypes.drop(1, axis=1, inplace=True)
cardsTypes.set_index(0)
cards['types'] = cardsTypes
cards.head()

colors convertedManaCost     layout   manaCost                   name  \
index                                                                         
0        [R]                 3     normal     {2}{R}         Abandon Reason   
1         []                 3     normal        {3}  Abandoned Sarcophagus   
2        [R]                 2     normal     {1}{R}    Abbot of Keral Keep   
3        [U]                 4  transform     {3}{U}    Aberrant Researcher   
4        [B]                 7     normal  {5}{B}{B}     Abhorrent Overlord   

              subtypes supertypes     types power toughness loyalty  
index                                                                
0                   []         []   Instant     0         0       0  
1                   []         []  Artifact     0         0       0  
2        [Human, Monk]         []  Creature     2         1       0  
3      [Human, Insect]         []  Creature     3         2       0  
4              [Demon]         []  Creature     6         6       0

In [23]:
# Trabalhando a coluna colors
pd.value_counts(cards['colors'])


TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[W]                989
[R]                968
[G]                960
[B]                960
[U]                953
[]                 851
[G, R]              81
[B, G]              80
[U, W]              80
[B, W]              80
[G, U]              79
[B, R]              77
[R, W]              74
[B, U]              74
[G, W]              74
[R, U]              73
[B, G, U]           10
[R, U, W]           10
[G, R, U]           10
[B, R, W]           10
[B, G, W]            9
[W, G]               7
[B, R, U]            7
[R, G]               5
[U, R]               5
[U, B]               5
[W, R]               5
[W, B]               5
[U, G]               5
[W, U]               5
[G, R, W]            4
[B, G, R, U, W]      4
[B, G, R]            3
[G, U, W]            3
[B, U, W]            2
Name: colors, dtype: int64

In [24]:
cardsColors = pd.DataFrame(cards['colors'])

In [25]:
# Iterar o dataframe para preencher os valores em formato de string
# Melhorar esse código - Talvez pensar em um método.
for (key,item) in cardsColors.iterrows():
    if (len(item['colors']) == 0):
        cardsColors.loc[key,'Cores'] = 'Colorless'
    elif (len(item['colors']) >= 2):
        cardsColors.loc[key,'Cores'] = 'Multicolor'
    elif (item['colors'].count('R') == 1):
        cardsColors.loc[key,'Cores'] = 'Red'
    elif (item['colors'].count('W') == 1):
        cardsColors.loc[key,'Cores'] = 'White'
    elif (item['colors'].count('G') == 1):
        cardsColors.loc[key,'Cores'] = 'Green'
    elif (item['colors'].count('U') == 1):
        cardsColors.loc[key,'Cores'] = 'Blue'
    elif (item['colors'].count('B') == 1):
        cardsColors.loc[key,'Cores'] = 'Black'
   

In [26]:
# Recuperando só o as cores
cardsColors.drop(columns='colors', inplace=True)

In [27]:
cardsColors

Cores
index            
0             Red
1       Colorless
2             Red
3            Blue
4           Black
...           ...
6562        Black
6563        Black
6564        Black
6565          Red
6566   Multicolor

[6567 rows x 1 columns]

In [28]:
# Concatenando a lista de Cores com o dataframe original
cards = pd.merge(cards, cardsColors, on='index')
# Deletando a coluna colors do dataframe original
cards.drop(columns='colors', inplace=True)
cards

convertedManaCost     layout      manaCost                   name  \
index                                                                     
0                     3     normal        {2}{R}         Abandon Reason   
1                     3     normal           {3}  Abandoned Sarcophagus   
2                     2     normal        {1}{R}    Abbot of Keral Keep   
3                     4  transform        {3}{U}    Aberrant Researcher   
4                     7     normal     {5}{B}{B}     Abhorrent Overlord   
...                 ...        ...           ...                    ...   
6562                  4     normal        {3}{B}              Zof Shade   
6563                  4     normal        {3}{B}     Zulaport Chainmage   
6564                  2     normal        {1}{B}     Zulaport Cutthroat   
6565                  1     normal           {R}      Zurgo Bellstriker   
6566                  5     normal  {2}{R}{W}{B}      Zurgo Helmsmasher   

                    subtypes   supertypes     types power toughness loyalty  \
index                                                                         
0                         []           []   Instant     0         0       0   
1                         []           []  Artifact     0         0       0   
2              [Human, Monk]           []  Creature     2         1       0   
3            [Human, Insect]           []  Creature     3         2       0   
4                    [Demon]           []  Creature     6         6       0   
...                      ...          ...       ...   ...       ...     ...   
6562                 [Shade]           []  Creature     2         2       0   
6563   [Human, Shaman, Ally]           []  Creature     4         2       0   
6564    [Human, Rogue, Ally]           []  Creature     1         1       0   
6565          [Orc, Warrior]  [Legendary]  Creature     2         2       0   
6566          [Orc, Warrior]  [Legendary]  Creature     7         2       0   

            Cores  
index              
0             Red  
1       Colorless  
2             Red  
3            Blue  
4           Black  
...           ...  
6562        Black  
6563        Black  
6564        Black  
6565          Red  
6566   Multicolor  

[6567 rows x 11 columns]

In [29]:
# Trabalhando a coluna subtypes
cards['subtypes']

index
0                          []
1                          []
2               [Human, Monk]
3             [Human, Insect]
4                     [Demon]
                ...          
6562                  [Shade]
6563    [Human, Shaman, Ally]
6564     [Human, Rogue, Ally]
6565           [Orc, Warrior]
6566           [Orc, Warrior]
Name: subtypes, Length: 6567, dtype: object

In [30]:
# Podemos ver que é importante manter todos os valores
# Para facilitar a manipulação e a visualização, transformaremos a lista em um dataframe
subtypes_df = cards['subtypes'].apply(pd.Series)
subtypes_df

0        1     2
index                      
0        NaN      NaN   NaN
1        NaN      NaN   NaN
2      Human     Monk   NaN
3      Human   Insect   NaN
4      Demon      NaN   NaN
...      ...      ...   ...
6562   Shade      NaN   NaN
6563   Human   Shaman  Ally
6564   Human    Rogue  Ally
6565     Orc  Warrior   NaN
6566     Orc  Warrior   NaN

[6567 rows x 3 columns]

In [31]:
# Tratando os valores nulos
subtypes_df.fillna("", inplace=True)
subtypes_df

0        1     2
index                      
0                          
1                          
2      Human     Monk      
3      Human   Insect      
4      Demon               
...      ...      ...   ...
6562   Shade               
6563   Human   Shaman  Ally
6564   Human    Rogue  Ally
6565     Orc  Warrior      
6566     Orc  Warrior      

[6567 rows x 3 columns]

In [32]:
# Renomeando o nome da colunas
colunas_subtypes = {0:'sybtypes1', 1:'subtypes2', 2:'subtypes3'}
subtypes_df.rename(columns = colunas_subtypes, inplace=True)
subtypes_df

sybtypes1 subtypes2 subtypes3
index                              
0                                  
1                                  
2         Human      Monk          
3         Human    Insect          
4         Demon                    
...         ...       ...       ...
6562      Shade                    
6563      Human    Shaman      Ally
6564      Human     Rogue      Ally
6565        Orc   Warrior          
6566        Orc   Warrior          

[6567 rows x 3 columns]

In [33]:
# Concatenando com o dataframe original
cards = pd.merge(cards, subtypes_df, on='index')
cards

convertedManaCost     layout      manaCost                   name  \
index                                                                     
0                     3     normal        {2}{R}         Abandon Reason   
1                     3     normal           {3}  Abandoned Sarcophagus   
2                     2     normal        {1}{R}    Abbot of Keral Keep   
3                     4  transform        {3}{U}    Aberrant Researcher   
4                     7     normal     {5}{B}{B}     Abhorrent Overlord   
...                 ...        ...           ...                    ...   
6562                  4     normal        {3}{B}              Zof Shade   
6563                  4     normal        {3}{B}     Zulaport Chainmage   
6564                  2     normal        {1}{B}     Zulaport Cutthroat   
6565                  1     normal           {R}      Zurgo Bellstriker   
6566                  5     normal  {2}{R}{W}{B}      Zurgo Helmsmasher   

                    subtypes   supertypes     types power toughness loyalty  \
index                                                                         
0                         []           []   Instant     0         0       0   
1                         []           []  Artifact     0         0       0   
2              [Human, Monk]           []  Creature     2         1       0   
3            [Human, Insect]           []  Creature     3         2       0   
4                    [Demon]           []  Creature     6         6       0   
...                      ...          ...       ...   ...       ...     ...   
6562                 [Shade]           []  Creature     2         2       0   
6563   [Human, Shaman, Ally]           []  Creature     4         2       0   
6564    [Human, Rogue, Ally]           []  Creature     1         1       0   
6565          [Orc, Warrior]  [Legendary]  Creature     2         2       0   
6566          [Orc, Warrior]  [Legendary]  Creature     7         2       0   

            Cores sybtypes1 subtypes2 subtypes3  
index                                            
0             Red                                
1       Colorless                                
2             Red     Human      Monk            
3            Blue     Human    Insect            
4           Black     Demon                      
...           ...       ...       ...       ...  
6562        Black     Shade                      
6563        Black     Human    Shaman      Ally  
6564        Black     Human     Rogue      Ally  
6565          Red       Orc   Warrior            
6566   Multicolor       Orc   Warrior            

[6567 rows x 14 columns]

In [34]:
# Removendo a antiga coluna sybtypes
cards.drop(columns='subtypes', axis=1, inplace=True)
cards

convertedManaCost     layout      manaCost                   name  \
index                                                                     
0                     3     normal        {2}{R}         Abandon Reason   
1                     3     normal           {3}  Abandoned Sarcophagus   
2                     2     normal        {1}{R}    Abbot of Keral Keep   
3                     4  transform        {3}{U}    Aberrant Researcher   
4                     7     normal     {5}{B}{B}     Abhorrent Overlord   
...                 ...        ...           ...                    ...   
6562                  4     normal        {3}{B}              Zof Shade   
6563                  4     normal        {3}{B}     Zulaport Chainmage   
6564                  2     normal        {1}{B}     Zulaport Cutthroat   
6565                  1     normal           {R}      Zurgo Bellstriker   
6566                  5     normal  {2}{R}{W}{B}      Zurgo Helmsmasher   

        supertypes     types power toughness loyalty       Cores sybtypes1  \
index                                                                        
0               []   Instant     0         0       0         Red             
1               []  Artifact     0         0       0   Colorless             
2               []  Creature     2         1       0         Red     Human   
3               []  Creature     3         2       0        Blue     Human   
4               []  Creature     6         6       0       Black     Demon   
...            ...       ...   ...       ...     ...         ...       ...   
6562            []  Creature     2         2       0       Black     Shade   
6563            []  Creature     4         2       0       Black     Human   
6564            []  Creature     1         1       0       Black     Human   
6565   [Legendary]  Creature     2         2       0         Red       Orc   
6566   [Legendary]  Creature     7         2       0  Multicolor       Orc   

      subtypes2 subtypes3  
index                      
0                          
1                          
2          Monk            
3        Insect            
4                          
...         ...       ...  
6562                       
6563     Shaman      Ally  
6564      Rogue      Ally  
6565    Warrior            
6566    Warrior            

[6567 rows x 13 columns]

In [35]:
# Trabalhando com a coluna supertypes
supertypes_df = cards['supertypes'].apply(pd.Series)
supertypes_df

0
index           
0            NaN
1            NaN
2            NaN
3            NaN
4            NaN
...          ...
6562         NaN
6563         NaN
6564         NaN
6565   Legendary
6566   Legendary

[6567 rows x 1 columns]

In [36]:
# Tratando os valores nulos e alterando o nome da coluna
supertypes_columns = {0:'supertypes'}
supertypes_df.rename(columns=supertypes_columns, inplace=True)
supertypes_df.fillna("Não se aplica",inplace=True)
supertypes_df

supertypes
index               
0      Não se aplica
1      Não se aplica
2      Não se aplica
3      Não se aplica
4      Não se aplica
...              ...
6562   Não se aplica
6563   Não se aplica
6564   Não se aplica
6565       Legendary
6566       Legendary

[6567 rows x 1 columns]

In [37]:
# Atribuindo os valores do dataframe do supertypes para o original
cards['supertypes'] = supertypes_df['supertypes']
cards

convertedManaCost     layout      manaCost                   name  \
index                                                                     
0                     3     normal        {2}{R}         Abandon Reason   
1                     3     normal           {3}  Abandoned Sarcophagus   
2                     2     normal        {1}{R}    Abbot of Keral Keep   
3                     4  transform        {3}{U}    Aberrant Researcher   
4                     7     normal     {5}{B}{B}     Abhorrent Overlord   
...                 ...        ...           ...                    ...   
6562                  4     normal        {3}{B}              Zof Shade   
6563                  4     normal        {3}{B}     Zulaport Chainmage   
6564                  2     normal        {1}{B}     Zulaport Cutthroat   
6565                  1     normal           {R}      Zurgo Bellstriker   
6566                  5     normal  {2}{R}{W}{B}      Zurgo Helmsmasher   

          supertypes     types power toughness loyalty       Cores sybtypes1  \
index                                                                          
0      Não se aplica   Instant     0         0       0         Red             
1      Não se aplica  Artifact     0         0       0   Colorless             
2      Não se aplica  Creature     2         1       0         Red     Human   
3      Não se aplica  Creature     3         2       0        Blue     Human   
4      Não se aplica  Creature     6         6       0       Black     Demon   
...              ...       ...   ...       ...     ...         ...       ...   
6562   Não se aplica  Creature     2         2       0       Black     Shade   
6563   Não se aplica  Creature     4         2       0       Black     Human   
6564   Não se aplica  Creature     1         1       0       Black     Human   
6565       Legendary  Creature     2         2       0         Red       Orc   
6566       Legendary  Creature     7         2       0  Multicolor       Orc   

      subtypes2 subtypes3  
index                      
0                          
1                          
2          Monk            
3        Insect            
4                          
...         ...       ...  
6562                       
6563     Shaman      Ally  
6564      Rogue      Ally  
6565    Warrior            
6566    Warrior            

[6567 rows x 13 columns]

## Agora temos um Dataframe tratado que podemos utilizar para gerar análises

In [38]:
# Cards separadas por Total de Mana Convertida
pd.value_counts(cards['convertedManaCost'], sort=False)

0.0      226
1.0      643
2.0     1421
3.0     1525
4.0     1180
5.0      788
6.0      489
7.0      197
8.0       54
9.0       24
10.0      10
11.0       6
12.0       3
13.0       1
Name: convertedManaCost, dtype: int64

- Podemos verificar que a maiora das cartas tem seu **custo total 3, seguidos de 2 e 4**
- Só existe **uma** carta com **custo 13** no formato Pionero do Magic The Gathering

In [39]:
# Cards seperados pelo Super Tipo
pd.value_counts(cards['supertypes'])

Não se aplica    6047
Legendary         514
Basic               6
Name: supertypes, dtype: int64

In [40]:
# PIVOT - Relacionando supertypes x types
cards.pivot_table('name', index=['supertypes','types'], aggfunc='count', margins=True)

name
supertypes    types             
Basic         Land             6
Legendary     Artifact        42
              Creature       260
              Enchantment     40
              Land            15
              Planeswalker   151
              Sorcery          6
Não se aplica Artifact       460
              Creature      3055
              Enchantment    617
              Instant        909
              Land           205
              Sorcery        801
All                         6567

In [41]:
cards.loc[(cards['supertypes'] == 'Basic')]

convertedManaCost  layout manaCost      name supertypes types power  \
index                                                                       
2127                  0  normal        0    Forest      Basic  Land     0   
2960                  0  normal        0    Island      Basic  Land     0   
3684                  0  normal        0  Mountain      Basic  Land     0   
4160                  0  normal        0    Plains      Basic  Land     0   
5579                  0  normal        0     Swamp      Basic  Land     0   
6349                  0  normal        0    Wastes      Basic  Land     0   

      toughness loyalty      Cores sybtypes1 subtypes2 subtypes3  
index                                                             
2127          0       0  Colorless    Forest                      
2960          0       0  Colorless    Island                      
3684          0       0  Colorless  Mountain                      
4160          0       0  Colorless    Plains                      
5579          0       0  Colorless     Swamp                      
6349          0       0  Colorless

- Temos um total de *514* cartas com supertipo Lendária
    - Podemos ver que das 514 a maioria das cartas são do **tipo Criatura, totalizando 260**.
- O supertipo **Basic** possui apenas 6 cartas e todas elas são do **tipo Terreno**
- **Só** temos cartas do tipo planeswalker com **supertipo lendária**